document/core/appendix/algorithm.rst

```diff
@@ -1,6 +1,193 @@
+In that case, an unknown type is returned.
+
+A second function for popping an operand takes an expected type, which the actual operand type is checked against.
+The types may differ in case one of them is Unknown.
+The more specific type is returned.
+
+Finally, there are accumulative functions for pushing or popping multiple operand types.
+
+The control stack is likewise manipulated through auxiliary functions:
+
+.. code-block:: none
+
+   func push_ctrl(label : list(val_type), out : list(val_type)) =
+     let frame = ctrl_frame(label, out, opds.size(), false)
+     ctrls.push(frame)
+
+   func pop_ctrl() : list(val_type) =
+     error_if(ctrls.is_empty())
+     let frame = ctrls.pop()
+     pop_opds(frame.end_types)
+     error_if(opds.size() =/= frame.height)
+     return frame.end_types
+
+   func unreachable() =
+     opds.resize(ctrl[0].height)
+     ctrl[0].unreachable := true
+
+Pushing a control frame takes the types of the label and result values.
+It allocates a new frame record recording them along with current height of the operand stack and marks the block as reachable.
+
+Popping a frame first checks that the control stack is not empty.
+It then verifies that the operand stack contains the right types of values expected at the end of the exited block and pops them off the operand stack.
+Afterwards, it checks that the stack has shrunk back to its initial height.
+
+Finally, the current frame can be marked as unreachable.
+In that case, all existing operand types are purged from the operand stack, in order to allow for the :ref:`stack-polymorphism <polymorphism>` logic in :code:`pop_opd` to take effect.
+
+.. note::
+   Even with the unreachable flag set, consecutive operands are still pushed to and popped from the operand stack.
+   That is necessary to detect invalid :ref:`examples <polymorphism>` like :math:`(\UNREACHABLE~(\I32.\CONST)~\I64.\ADD)`.
+   However, a polymorphic stack cannot underflow, but instead generates :code:`Unknown` types as needed.
+
+
+.. index:: opcode
+
+Validation of Opcode Sequences
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following function shows the validation of a number of representative instructions that manipulate the stack.
+Other instructions are checked in a similar manner.
+
+.. note::
+   Various instructions not shown here will additionally require the presence of a validation :ref:`context <context>` for checking uses of :ref:`indices <syntax-index>`.
+   That is an easy addition and therefor omitted from this presentation.
+
+.. code-block:: none
+
+   func validate(opcode) =
+     switch (opcode)
+       case (i32.add)
+         pop_opd(I32)
+         pop_opd(I32)
+         push_opd(I32)
+
+       case (drop)
+         pop_opd()
+
+       case (select)
+         pop_opd(I32)
+         let t1 = pop_opd()
+         let t2 = pop_opd(t1)
+         push_opd(t2)
+
+       case (unreachable)
+         unreachable()
```

Comment:
Hm, `return` requires the context, which is why I left it out. It wasn't actually the intention to cover all uses of unreachable. You think it is easy enough to extrapolate that from the rules for branches?